### PR TITLE
Add a method for setting `Camera.view_direction` and `Camera.up_direction`

### PR DESCRIPTION
**Pull request by alisterburt**
# Description
This PR adds a setter on the `view_direction` attribute of our `Camera` model. This is useful because we often have a vector describing a direction in our data and might not know how to convert this to the Euler angle representation used by our camera model.

https://user-images.githubusercontent.com/7307488/140627857-c5171135-1858-4433-9f5a-2bee1ee42ecd.mp4

The view direction alone does not uniquely describe the camera orientation. I have provided sane defaults which I think match expectations if only a view direction is provided. I have also added the slightly more verbose method `Camera.set_view_direction(view_direction, up_direction)` which additionally specifies a vector which should point upwards on the canvas. These two attributes uniquely define the camera orientation.

cc @kevinyamauchi @brisvag 

Also cc @tlambert03 and @andy-sweet because I had to override `__setattr__` on the `Camera` to achieve the `Camera.view_direction = ` API [as discussed on zulip](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/pydantic.20model.20with.20settable.20property) and you might find this icky 🤮

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?
- [x] simple tests added to test suite

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/components/camera.py
Comment (alisterburt) [inline napari/components/camera.py:17]: nitpick unrelated to this PR - can separate out if anyone cares!

File: napari/components/camera.py
Comment (sofroniewn) [inline napari/components/camera.py:73]: I think this should be private, otherwise it's a bit redundant with the `view_direction` setter. Can it also get a more complete doc string

File: napari/components/camera.py
Comment (sofroniewn) [inline napari/components/camera.py]: I'll leave a comment on the main thread that I don't love overwriting `__setattr__` - but if we are going to go with this approach then I'd like us to make something like `SettableEventedModel` that lives alongside our `EventedModel` and gets used here. I feel like this pattern is far to generic to live on the camera and if we like it here we might want to start using it elsewhere. If we really like it, then it could just live on `EventedModel` directly - thoughts @tlambert03?

File: napari/components/camera.py
Comment (sofroniewn) [reply napari/components/camera.py:17]: fine with me

**Review by sofroniewn — COMMENTED** [review]
I know this was discussed at length over zulip, but I'm looking at the PR now and am wondering if this is the best approach. Overwriting `__setattr__` really feels like we're fighting against pydantic here.

Is there no way to achieve something like this with validators instead, even a root validator? I guess the tricky thing is if we put `view_direction` and `angles` both in the pydantic model they will both emit events. Maybe that's ok though and we can make this work. We'd have to use a root validator and when one changes update the other one. An upside is we only do calculations once during setting, never during getting - which is good if we'll do more getting than setting. It's a downside if we do more setting. Thoughts @alisterburt @tlambert03?

**Review by kevinyamauchi — COMMENTED** [review]
Thanks for the PR, @alisterburt ! Being able to set view_direction more easily with definitely make my work easier. 

> Overwriting __setattr__ really feels like we're fighting against pydantic here.

I agree with this. I am not fully caught up on the zulip thread, so please disregard if these points were already discussed, but I've listed a few thoughts below

- I like having the `view_direction` property and It think a lot of people will largely interact with that rather than the angles.
- Perhaps semantics, but I am not a huge fan of creating a `view_direction` setter if the model uses `angles`. It makes the model feel overdefined for me.
- What if instead of implementing `Camera.set_view_direction`  as a `Camera.view_direction` setter, we implement it as a method explicitly for setting angles in a convenient way (e.g.,  `Camera.set_angles_from_view_direction()`)?
- An alternative is to make the `view_direction` and `up_direction` the model fields and then `angles` a derived property (I think @tlambert03 suggested this)
- We could handle the conversion in the validator if we specify that setting `Camera.angles` with a view direction requires both a view_direction and an up_direction (i.e.., a (2, 3) array), which would be distinguishable from setting via euler angles. I am not a fan of this though because I think it's a bit too magic (i.e., confusing)


edit: an argument for keeping the model field as `Camera.angles` is that this is what the vispy camera uses and the `Camera` model gets updated everything the camera changes, so we would be doing the conversion a lot in the background if we change the model field to `Camera.view_direction` and `Camera.up_direction`.

File: napari/components/camera.py
Comment (kevinyamauchi) [inline napari/components/camera.py]: It might be handy to make the default up direction the current up direction. This would make it easier to rotate around an object.

**Comment by kevinyamauchi** [comment]
Sorry, one more thing: while we're at it, I think it would be nice to add an `up_direction` property that is derived from the angles.

**Comment by tlambert03** [comment]
I think we should probably more strongly consider the `set_view_direction()` option. It is a lot of hoops to jump through, and using a setter method would make it obvious that this wasn't a core model attribute, but rather a convenience.

**Comment by alisterburt** [comment]
Thanks for all of your input @sofroniewn @kevinyamauchi and @tlambert03 ! 

I think the key things that we want from this PR are
1. the ability to set just view direction
2. the ability to set view direction and up direction
3. minimal redundancy in the model
4. not to clutter the camera model with generic stuff

I think if we only wanted 1 and not 2,3,4 then a setter would have been the correct API but given that we have 2,3,4 I'm inclined to agree with @tlambert03 and @kevinyamauchi 's `Camera.set_view_direction()` suggestion - I much prefer this to trying to keep multiple attributes on the model in sync 🙂 

> An alternative is to make the view_direction and up_direction the model fields and then angles a derived property (I think @tlambert03 suggested this)

I definitely think we still want/need the angles exposed, they're a much more convenient interface for simple rotations around x/y/z. If we were to change the model I think it should be a rotation matrix from which everything else is derived but I don't think rocking the boat this much is necessary.

I will remove the setter and provide only the `set_view_direction` API. 

@kevinyamauchi I like your suggestion of keeping the current up direction by default but am not sure of the best way to handle that if someone provides the current up direction as the new view direction -> I will add the `Camera.up_direction` property so that anyone can easily use this in their call to `Camera.set_view_direction()`. This has the benefit of providing predictable behaviour if someone habitually just sets the view direction.

Will update the API and the docstring based on suggestions now, thanks again!!

**Comment by sofroniewn** [comment]
> I will remove the setter and provide only the set_view_direction API.
> ....
> Will update the API and the docstring based on suggestions now, thanks again!!

+1 from me, good discussion here 🚀

File: napari/components/camera.py
Comment (alisterburt) [reply napari/components/camera.py:73]: this has become the only API and others were remove, keeping it public because of that

**Comment by alisterburt** [comment]
Cheers @sofroniewn ! I've now addressed all comments, added an `up_direction` property and assuming this passes CI it should be good to go.

I'm going to leave this script here too, figuring out how to set our camera angles properly to achieve this wasn't trivial and it might help someone in the future 🙂 

```python
import numpy as np
from napari.components import Camera
from scipy.spatial.transform import Rotation as R

camera = Camera(angles=(1, 2, 3))
original_eulers = camera.angles
original_view_dir = np.array(camera.view_direction)

for seq in ('xyz', 'yzx', 'zxy', 'xzy', 'zyx', 'yxz'):
    for s in (seq, seq.upper()):
        matrix = R.from_euler(s, original_eulers, degrees=True).as_matrix()
        view_dir = original_view_dir[::-1]
        for i in range(3):
                if np.allclose(matrix[i, :], view_dir):
                    print(s, 'row', i)
                    print(s, view_dir, '\n', matrix, '\n\n\n')
                elif np.allclose(matrix[:, i], view_dir):
                    print(s, 'column', i)
                    print(s, view_dir, '\n', matrix, '\n\n\n')

matrix = R.from_euler(seq='yzx', angles=original_eulers, degrees=True).as_matrix()
print(matrix)
angles = R.from_matrix(matrix).as_euler(seq='yzx', degrees=True)

print(angles, original_eulers)

# yzx gives a matrix which we can construct...
# construct matrix
up_dir = np.cross(view_dir, (1, 1, 1))
up_dir /= np.linalg.norm(up_dir)
third = np.cross(up_dir, view_dir)
third /= np.linalg.norm(third)

matrix = np.column_stack((up_dir, view_dir, third))
print(matrix)

eulers = R.from_matrix(matrix).as_euler(seq='yzx', degrees=True)

camera.angles = eulers
print(original_view_dir, camera.view_direction)
```

**Comment by alisterburt** [comment]
min_req tests were because I used `scipy.spatial.transform.Rotation.as_matrix()` which was [renamed from `Rotation.as_dcm()` in scipy 1.4.0](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.as_matrix.html) which was released in December 2019

I've bumped the version, assuming that will solve issues and people are okay with the bump!

**Review by sofroniewn — APPROVED** [review]
Love it, LGTM! so much cleaner code, will be much simpler to understand and maintain.

**Review by kevinyamauchi — APPROVED** [review]
Looks great @alisterburt ! I will use this as soon as it's merged :)

**Comment by alisterburt** [comment]
Thanks for the quick turnaround on this all of you, pleasure as always 🙂

**Comment by sofroniewn** [comment]
I'm going to merge this now so development can press on, it's gotten two approves and matches the request of a third reviewer, so i think we're good to go - but if anyone wants any revisions before release we can make them. thanks all!!

**Pull request merged**
